src/boot/lib/tensor.ml

```diff
@@ -0,0 +1,199 @@
+let tensor_shape_and_index_does_not_match =
+  "Tensor shape and index does not match"
+
+let tensor_shape_mismatch = "Tensor shape mismatch"
+
+let tensor_shape_and_ofs_and_len_does_not_match =
+  "Tensor shape and ofs and len does not match"
+
+let prod = Array.fold_left ( * ) 1
+
+let sum = Array.fold_left ( + ) 0
+
+let row_major_ofs shape is =
+  let n = Array.length shape in
+  let d = Array.length is in
+  let tmp_ofs = ref 0 in
+  let tmp = ref 0 in
+  for k = 0 to d - 1 do
+    tmp := 1 ;
+    for l = k + 1 to n - 1 do
+      tmp := !tmp * shape.(l)
+    done ;
+    tmp_ofs := !tmp_ofs + (!tmp * is.(k))
+  done ;
+  !tmp_ofs
+
+let inverse_row_major_ofs shape i =
+  let d = Array.length shape in
+  let is = Array.make d 0 in
+  let ofs = ref i in
+  for k = 0 to d - 1 do
+    let l = d - 1 - k in
+    let n = shape.(l) in
+    let j = !ofs mod n in
+    is.(l) <- j ;
+    ofs := !ofs / n
+  done ;
+  is
+
+let mk_iteri rank shape slice f t =
+  if rank t = 0 then f (-1) t
+  else
+    let n = (shape t).(0) in
+    for i = 0 to n - 1 do
+      f i (slice t [|i|])
+    done
+
+module NoNum = struct
+  type 'a t =
+    {data: 'a array; shape: int array; rank: int; left_ofs: int; size: int}
+
+  let rank t = t.rank
+
+  let shape t = t.shape
+
+  let create shape f =
+    let rank = Array.length shape in
+    let size = prod shape in
+    let data = Array.init size (fun i -> f (inverse_row_major_ofs shape i)) in
+    let left_ofs = 0 in
+    {data; rank; shape; left_ofs; size}
+
+  let is_valid_index shape is =
+    Array.length shape = Array.length is
+    && Array.for_all2 (fun n i -> i >= 0 && i < n) shape is
```

Comment:
The function `Array.for_all2` requires OCaml version 4.11.0, which is newer than the most recen one provided by some package managers (see https://ocaml.org/docs/install.html). I thought it could be good to point this out since I could not find a version requirement of OCaml in the README.

Reply:
Good to know, then I'll change this implementation.

Reply:
Since there are no guidelines, it's hard to know where to draw the line. Also, even though 4.11.0 isn't available from all package managers, it could be installed via `opam` as described [here](https://stackoverflow.com/a/40901155). I think a minimum supported OCaml version could be a good discussion point in a future meeting.